### PR TITLE
alignments.json save interval for extracting

### PR DESCRIPTION
**Pull request by Tvde1**
Added an optional parameter `-si` or `--save-interval` which will save the alignments.json after the set amount of frames.

So `python faceswap.py extract ... -si 5` will save the file after 5 frames.
It will also skip the already extracted files when it's restarted.

### CODE REVIEW

File: lib/cli.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import os
 import platform
 import sys
+from math import inf
```
Comment (torzdf) [inline lib/cli.py]: Best not to have this import as it's not necessary. Make the default `None `and handle saving if a value is passed

File: lib/cli.py [hunk 2]
```diff
@@ -427,6 +428,16 @@ def get_optional_arguments():
                               "help": "Perform extra alignment to ensure "
                                       "left/right eyes are  at the same "
                                       "height"})
+        argument_list.append({"opts": ("-si", "--save-interval"),
+                              "dest": "save_interval",
+                              "type": int,
+                              "default": inf,
```
Comment (torzdf) [inline lib/cli.py]: Change to None

File: scripts/extract.py [hunk 3]
```diff
@@ -44,19 +45,33 @@ def process(self):
...
         self.faces.num_faces_detected = faces
 
+    def write_alignments(self):
+        self.alignments.write_alignments(self.faces.faces_detected)
+
+
     def extract_single_process(self):
         """ Run extraction in a single process """
+
+        alignments = self.alignments.read_alignments()
```
Comment (torzdf) [inline scripts/extract.py]: Why are you reading alignments here? It shouldn't exist at this point. If it's to continue off from where previously left off, then we have 'skip existing' for that.

File: scripts/extract.py [hunk 4]
```diff
@@ -44,19 +45,33 @@ def process(self):
...
+        
+        frame_no = 0
         for filename in tqdm(self.images.input_images, file=sys.stdout):
+            basename = os.path.basename(filename)
+            if basename in alignments:
```
Comment (torzdf) [inline scripts/extract.py]: See above. This looks like it's duplicating skip-existing code.

File: scripts/extract.py [hunk 5]
```diff
@@ -44,19 +45,33 @@ def process(self):
...
+                continue
             filename, faces = self.process_single_image(filename)
-            self.faces.faces_detected[os.path.basename(filename)] = faces
+            self.faces.faces_detected[basename] = faces
+            frame_no = frame_no + 1
```
Comment (torzdf) [inline scripts/extract.py]: `frame_no += 1`

File: scripts/extract.py [hunk 6]
```diff
@@ -44,19 +45,33 @@ def process(self):
...
+            if frame_no == self.save_interval:
```
Comment (torzdf) [inline scripts/extract.py]: `if hasattr(self.args, "save_interval") and self.args.save_interval == frame_no`

**Comment by Tvde1** [comment]
Thanks for the feedback. I've changed the stuff you requested

File: scripts/extract.py [hunk 7]
```diff
@@ -27,6 +27,7 @@ def __init__(self, arguments):
         self.output_dir = self.faces.output_dir
 
         self.export_face = True
+        self.save_interval = self.args.save_interval
```
Comment (torzdf) [inline scripts/extract.py]: If extraction is being run as part of convert (i.e. no alignments file exists when someone runs convert) then `save_interval` won't exist in `arguments` and the code will error out, so remove this line and see below.

File: scripts/extract.py [hunk 8]
```diff
@@ -44,22 +45,31 @@ def process(self):
...
+
+
     def extract_single_process(self):
         """ Run extraction in a single process """
+        frame_no = 0
         for filename in tqdm(self.images.input_images, file=sys.stdout):
             filename, faces = self.process_single_image(filename)
             self.faces.faces_detected[os.path.basename(filename)] = faces
+            frame_no += 1
+            if frame_no == self.save_interval:
```
Comment (torzdf) [inline scripts/extract.py]: As per earlier comment, self.save_interval will not necessarily exist, so reference it directly from arguments and make sure it is a passed argument: `if hasattr(self.args, "save_interval") and self.args.save_interval == frame_no`

File: scripts/extract.py [hunk 9]
```diff
@@ -68,6 +78,10 @@ def extract_multi_process(self):
                 file=sys.stdout):
             self.faces.num_faces_detected += 1
             self.faces.faces_detected[os.path.basename(filename)] = faces
+            frame_no += 1
+            if frame_no == self.save_interval:
```
Comment (torzdf) [inline scripts/extract.py]: As per earlier comment, self.save_interval will not necessarily exist, so reference it directly from arguments and make sure it is a passed argument: `if hasattr(self.args, "save_interval") and self.args.save_interval == frame_no`

File: scripts/extract.py (hunk 7 above)
Comment (Tvde1) [reply scripts/extract.py]: I added the check here so we don't have to keep looking if the arg exists. I don't know if it matters much with performance but it's a tiny bit optimization I guess

**Pull request merged**